### PR TITLE
Clean-up and simplify Tracing-Options processing

### PR DESCRIPTION
**Pull request by HannesWell**
@fedejeanne these are the clean-ups I mentioned in https://github.com/eclipse-pde/eclipse.pde/pull/674#pullrequestreview-1577719920.
Since you are probably using tracing more often than I do, do you want to try/review the changes?

### CODE REVIEW

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java
Comment (HannesWell) [inline ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java:87]: From this discussion I assume, that this is a save cast:
https://stackoverflow.com/q/873510/14542697

File: ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/launcher/LaunchArgumentsHelper.java
Comment (fedejeanne) [inline ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/launcher/LaunchArgumentsHelper.java:284]: I'd simplify it (remove `COMMA`) and I'd also trim the elements inside this method, just to make sure the white spaces don't cause any obscure trouble in the future:

```java
public static Stream<String> splitElementsByComma(String value) {
      return Arrays.stream(value.split(",")).map(String::trim); //$NON-NLS-1$
}
```

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java
Comment (fedejeanne) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java:220]: If it's not a field anymore then you should rename it to `descriptors`

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java
Comment (fedejeanne) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java:241]: FYI `Boolean.parseBoolean` uses `equalsIgnoreCase` so this change is **not** a refactoring and this line is now **not** in sync with the rest of the the class (_e.g._ with the ones in line 238). 

I would either revert this change and keep it consistent with the rest of the logic in the class (**case sensitive** _i.e._ "true" is `true` but "TrUe" is `false`) or I would adapt the logic in the other places and let the whole class work in a **case insensitive** manner (_i.e._ "true" is `true` and "TrUe" is also `true`). 

Your call :-)

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java
Comment (fedejeanne) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java:265]: I would extract 2 new private methods:
* `intToBoolean`
* `booleanToInt`

You could then make these 2 transformations consistent to each other and that would also improve readability and avoid obscure errors in the future (a typical one would be that someone flips this `if`-logic and suddenly `0 == true` in some other part of the code)

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java
Comment (fedejeanne) [inline ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java:200]: You could change `"\n"` with the system-independent version: `System.lineSeparator()`

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java
Comment (HannesWell) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java:87]: Thanks for verifying this!

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java
Comment (HannesWell) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java:197]: `Properties.store()` states in its java-doc that `The stream is written using the ISO 8859-1 character encoding.` and `Properties.load()` assume the same encoding. But if the InputStream is processed by a default Reader, that encoding has to be specified explicitly.

File: ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/launcher/LaunchArgumentsHelper.java
Comment (HannesWell) [reply ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/launcher/LaunchArgumentsHelper.java:284]: Using trim directly is a good suggestion, but I'm not so convinced of using `value.split(",")` since this creates an array of all resulting elements of the split, while using `Pattern.splitAsStream()` creates a lazy stream of elements and therefore does not need the intermediate array and could terminate before reaching the end.
Of course it is definitely questionable if there is a real performance and memory usage gain from this small optimization (probably at least not a relevant one).

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java
Comment (HannesWell) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java:220]: Good point.
But looking at it again, the list is only populated but never read, so effectively unused and can be removed.

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java
Comment (HannesWell) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java:241]: Thanks for this remark, that's a good point as well.
But looking at the vicinity of this line, especially line 238 it contains `lvalue.equals("true") || lvalue.equals("false")` and `lvalue` is `lvalue = value.toLowerCase(Locale.ENGLISH)`, so at least that is already comparing not case-sensitive, just in a more complicated way (btw. using toLowerCase seems to be the pattern used before `equalsIgnoreCase()` case existed). But using `equalsIgnoreCase` allows to remove `lvalue` as well and operate on `value` directly.

In general I wonder why boolean values are stored as Integer values in the fValues Map<String,Object> while it could simply use Booleans?
Just using Booleans would also simplify the save() method below because the Boolean could be just converted into a String and there would be no need to convert the int to a `true` or `false` String again.

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java
Comment (HannesWell) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java:200]: The comment string is later split by `\\r?\\n` in the only caller in `TracingPropertySource.getFormattedComment()`, so this does not really make a difference so I would therefore just keep it as is to make it simpler.

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java
Comment (HannesWell) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java:265]: As commented above I think this overall story is just simplified if Boolean objects are stored in the map instead of mapping boolean values to Integer/int values. Now clue why this choice was made before, but in the current form of the code I see no reason why not to use Boolean objects in stead.
Then all this is not necessary.

**Comment by HannesWell** [comment]
@fedejeanne I pushed an enhanced commit that mainly changes the handling of boolean values in `TracingPropertySource`.

If possible, another view is more than welcome, but I would like to to submit this tomorrow evening in order to have at least this in M1.

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java:197]: Oh, I missed that, thanks for pointing it out! It makes sense then to specify the charset. I hope this doesn't bring problems in the future if someone edits the file with a text editor and (inadvertently) changes the charset.

I think some implementation detail like that would be worth being documented in a comment, something like `// the charset needs to be set explicitly because Properties.store() also uses this charset per default`

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java]: > But looking at the vicinity of this line, especially line 238...

👍 

> In general I wonder why boolean values are stored as Integer values in the fValues Map<String,Object> while it could simply use Booleans?

`0`s and `1`s were probably allowed in the configuration file before and since the method `getOptions(IPluginModelBase)` doesn't convert them to "real" `boolean`s then the logic had to be moved someplace else. If that's the case then removing the logic should be considered a breaking change.

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java]: See my previous comment about this probably being a **breaking change**

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java]: On the plus side: I don't see anywhere in the documentation that 0s and 1s are allowed to be used as booleans and I don't see anywhere that one writes 0s and 1s into the file so it should be fine to delete it.

Or do you see any place were 0s and 1s are written into the file?

**Comment by fedejeanne** [comment]
# ⚠️ WARNING ⚠️ 

I just tested the changes and found an error.

## Steps to reproduce
1. Launch a product --> a _Run configuration_ will be created
2. Open the _Run configurations_ dialog and select the newly created PDE configuration
3. Switch to the _Tracing_ tab and **Enable tracing** (check it)
4. Select a plugin and click **Apply**
5. Select another plugin and click **Apply** again

This dialog comes up and asks you if you want to overwrite the previous configuration:
image

I tried clicking....:
* **_Save_** --> the dialog keeps coming up
* **_Don't save_**: I get a `SAXParseException` (see below) and then my configuration is deleted
* **_Cancel_**: I get the same `SAXParseException` (see below) and then the dialog pops up again

Here's the exception: 

```java
org.eclipse.debug.core.DebugException: org.xml.sax.SAXParseException; lineNumber: 39; columnNumber: 28; Zeichenreferenz "&#0" ist ein ungültiges XML-Zeichen. occurred while reading launch configuration file: D:\dev\eclipse-contributors-experimental\runtime-sdk.product\.metadata\.plugins\org.eclipse.debug.core\.launches\aquintosEE.product.launch.
	at org.eclipse.debug.internal.core.LaunchManager.createDebugException(LaunchManager.java:777)
	at org.eclipse.debug.internal.core.LaunchManager.throwException(LaunchManager.java:2304)
	at org.eclipse.debug.internal.core.LaunchManager.getInfo(LaunchManager.java:1243)
	at org.eclipse.debug.internal.core.LaunchConfiguration.getInfo(LaunchConfiguration.java:487)
	at org.eclipse.debug.internal.core.LaunchConfigurationWorkingCopy.copyFrom(LaunchConfigurationWorkingCopy.java:477)
	at org.eclipse.debug.internal.core.LaunchConfigurationWorkingCopy.<init>(LaunchConfigurationWorkingCopy.java:101)
	at org.eclipse.debug.internal.core.LaunchConfiguration.getWorkingCopy(LaunchConfiguration.java:648)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationTabGroupViewer.handleRevertPressed(LaunchConfigurationTabGroupViewer.java:1545)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationsDialog.handleLaunchConfigurationSelectionChanged(LaunchConfigurationsDialog.java:1042)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationsDialog.lambda$1(LaunchConfigurationsDialog.java:614)
	at org.eclipse.jface.viewers.StructuredViewer$3.run(StructuredViewer.java:827)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
	at org.eclipse.jface.viewers.StructuredViewer.firePostSelectionChanged(StructuredViewer.java:824)
	at org.eclipse.jface.viewers.StructuredViewer.setSelection(StructuredViewer.java:1676)
	at org.eclipse.jface.viewers.TreeViewer.setSelection(TreeViewer.java:1106)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationViewer.preservingSelection(LaunchConfigurationViewer.java:120)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1474)
	at org.eclipse.jface.viewers.ColumnViewer.refresh(ColumnViewer.java:523)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1428)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationsDialog.refreshLaunchConfigurationView(LaunchConfigurationsDialog.java:1344)
	at org.eclipse.debug.ui.PrototypeTab.postApply(PrototypeTab.java:486)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationsDialog.handleLaunchConfigurationSelectionChanged(LaunchConfigurationsDialog.java:1036)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationsDialog.lambda$1(LaunchConfigurationsDialog.java:614)
	at org.eclipse.jface.viewers.StructuredViewer$3.run(StructuredViewer.java:827)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
	at org.eclipse.jface.viewers.StructuredViewer.firePostSelectionChanged(StructuredViewer.java:824)
	at org.eclipse.jface.viewers.StructuredViewer.setSelection(StructuredViewer.java:1676)
	at org.eclipse.jface.viewers.TreeViewer.setSelection(TreeViewer.java:1106)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationViewer.preservingSelection(LaunchConfigurationViewer.java:120)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1474)
	at org.eclipse.jface.viewers.ColumnViewer.refresh(ColumnViewer.java:523)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1428)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationsDialog.refreshLaunchConfigurationView(LaunchConfigurationsDialog.java:1344)
	at org.eclipse.debug.ui.PrototypeTab.postApply(PrototypeTab.java:486)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationTabGroupViewer$4.widgetSelected(LaunchConfigurationTabGroupViewer.java:342)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4274)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4072)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3660)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchConfigurationsDialog.open(LaunchConfigurationsDialog.java:1240)
	at org.eclipse.debug.ui.DebugUITools.lambda$1(DebugUITools.java:630)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:67)
	at org.eclipse.debug.ui.DebugUITools.openLaunchConfigurationDialogOnGroup(DebugUITools.java:636)
	at org.eclipse.debug.ui.DebugUITools.openLaunchConfigurationDialogOnGroup(DebugUITools.java:574)
	at org.eclipse.debug.ui.actions.OpenLaunchDialogAction.run(OpenLaunchDialogAction.java:85)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4274)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4072)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3660)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:342)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:645)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:342)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:552)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:651)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:588)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1459)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1432)
Caused by: org.xml.sax.SAXParseException; lineNumber: 39; columnNumber: 28; Zeichenreferenz "&#0" ist ein ungültiges XML-Zeichen.
	at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:204)
	at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:178)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:400)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1465)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLScanner.scanCharReferenceValue(XMLScanner.java:1338)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLScanner.scanAttributeValue(XMLScanner.java:895)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanAttribute(XMLDocumentFragmentScannerImpl.java:1522)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanStartElement(XMLDocumentFragmentScannerImpl.java:1363)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2726)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:247)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:342)
	at org.eclipse.debug.internal.core.LaunchManager.createInfoFromXML(LaunchManager.java:827)
	at org.eclipse.debug.internal.core.LaunchManager.getInfo(LaunchManager.java:1236)
	... 78 more
```

I reverted your changes (back to _master_) and the error does **not** occur.

Could you please look into it before merging!

**Review by fedejeanne — APPROVED** [review]
I just rebased on master and tested again and the error does **not** occur anymore. This PR LGTM

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java
Comment (HannesWell) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingPropertySource.java]: > Or do you see any place were 0s and 1s are written into the file?

Nope. And from the code I don't think it is even possible, so this change seems save.

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java
Comment (HannesWell) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java:197]: > Oh, I missed that, thanks for pointing it out! It makes sense then to specify the charset. I hope this doesn't bring problems in the future if someone edits the file with a text editor and (inadvertently) changes the charset.

Then that's the fault of the person, but nothing we can do about. :/


> I think some implementation detail like that would be worth being documented in a comment, something like `// the charset needs to be set explicitly because Properties.store() also uses this charset per default`

Good point, done that.

**Comment by HannesWell** [comment]
> I just rebased on master and tested again and the error does **not** occur anymore. This PR LGTM

Thanks for verifying this again and your reviews in general. :)

After fixing the regression discovered by the tests (thanks Julian that we have them) and applying a few more small clean ups this is now ready.

**Comment by fedejeanne** [comment]
> ... (thanks Julian that we have them) ...

Will do 😄 

Thank you too for merging! I'll try to come back to my original issue this week so we can improve the user experience when opening the PDE _Run configuration_.

**Pull request merged**
